Missing nullability warnings about member initialization in constructors in structs with default struct ctor calls

**Version Used**: 
```
Branch master (13 Oct 2020)
Latest commit 941dc7e by Shimmy:
Default constructor suggestion between members (#48318) (#48503)
```

**Steps to Reproduce**:

Compile and run the following code
```cs
#nullable enable

new S(1);

struct S
{
  string f;
  public S(long x) : this() => f.ToString();
}
```

**Expected Behavior**:
`warning CS8602: Dereference of a possibly null reference.` for `f.ToString()`

**Actual Behavior**:
No nullability warnings. The code crashes at runtime with a `NullReferenceException`

**Notes**
Roslyn seems to think that `: this()` constructor call in struct correctly initializes all struct fields which it doesn't. Roslyn should just ignore default stuct constructor calls when enforcing nullability contract of a constructor.

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -444,7 +444,6 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
             this.Diagnostics.Clear();
             this.regionPlace = RegionPlace.Before;
-            makeNotNullMembersMaybeNull();
             if (!_isSpeculative)
             {
                 ParameterSymbol methodThisParameter = MethodThisParameter;
@@ -454,6 +453,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     EnterParameter(methodThisParameter, methodThisParameter.TypeWithAnnotations);
                 }
 
+                makeNotNullMembersMaybeNull();
                 // We need to create a snapshot even of the first node, because we want to have the state of the initial parameters.
                 _snapshotBuilderOpt?.TakeIncrementalSnapshot(methodMainNode, State);
             }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs
@@ -825,7 +825,7 @@ struct S
         }
 
         [Fact]
-        public void StructConstructorInitializer()
+        public void StructConstructorInitializer_UninitializedProperty()
         {
             var source = @"
 struct S1
@@ -866,6 +866,54 @@ struct S1
                 // (15,12): warning CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.
                 //     public S1(object obj1, object obj2) : this() // 4
                 Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "S1").WithArguments("property", "Prop").WithLocation(15, 12));
+        }
+
+        [Fact, WorkItem(48574, "https://github.com/dotnet/roslyn/issues/48574")]
+        public void StructConstructorInitializer_UninitializedField()
+        {
+            var source = @"
+struct S1
+{
+    public string field; // 0
+    public S1(string s) // 1
+    {
+        field.ToString(); // 2
+    }
+
+    public S1(object obj) : this()
+    {
+        field.ToString(); // 3
+    }
+
+    public S1(object obj1, object obj2) : this() // 4
+    {
+    }
+
+    public S1(string s1, string s2) : this(s1)
+    {
+        field.ToString();
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (4,19): warning CS0649: Field 'S1.field' is never assigned to, and will always have its default value null
+                //     public string field; // 0
+                Diagnostic(ErrorCode.WRN_UnassignedInternalField, "field").WithArguments("S1.field", "null").WithLocation(4, 19),
+                // (5,12): error CS0171: Field 'S1.field' must be fully assigned before control is returned to the caller
+                //     public S1(string s) // 1
+                Diagnostic(ErrorCode.ERR_UnassignedThis, "S1").WithArguments("S1.field").WithLocation(5, 12),
+                // (7,9): error CS0170: Use of possibly unassigned field 'field'
+                //         field.ToString(); // 2
+                Diagnostic(ErrorCode.ERR_UseDefViolationField, "field").WithArguments("field").WithLocation(7, 9),
+                // (12,9): warning CS8602: Dereference of a possibly null reference.
+                //         field.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field").WithLocation(12, 9),
+                // (15,12): warning CS8618: Non-nullable field 'field' must contain a non-null value when exiting constructor. Consider declaring the field as nullable.
+                //     public S1(object obj1, object obj2) : this() // 4
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "S1").WithArguments("field", "field").WithLocation(15, 12)
+                );
         }
 
         [Fact, WorkItem(43215, "https://github.com/dotnet/roslyn/issues/43215")]

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs
@@ -1534,6 +1534,62 @@ class C
             }
         }
 
+        [Fact, WorkItem(48574, "https://github.com/dotnet/roslyn/issues/48574")]
+        public void SpeculativeGetTypeInfo_Constructor()
+        {
+            var source = @"
+class C
+{
+    public string Prop { get; set; }
+    public C()
+    {
+        if (Prop != null)
+        {
+            Prop.ToString();
+        }
+
+        Prop?.ToString();
+
+        Prop = """";
+        var s2 = Prop == null ? """" : Prop;
+    }
+}";
+
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (5,12): warning CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.
+                //     public C()
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("property", "Prop").WithLocation(5, 12));
+
+            var syntaxTree = comp.SyntaxTrees[0];
+            var root = syntaxTree.GetRoot();
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            var ifStatement = root.DescendantNodes().OfType<IfStatementSyntax>().Single();
+            var conditionalAccessExpression = root.DescendantNodes().OfType<ConditionalAccessExpressionSyntax>().Single();
+            var ternary = root.DescendantNodes().OfType<ConditionalExpressionSyntax>().Single();
+
+            var newReference = (IdentifierNameSyntax)SyntaxFactory.ParseExpression(@"Prop");
+            var newCoalesce = (AssignmentExpressionSyntax)SyntaxFactory.ParseExpression(@"Prop ??= """"");
+
+            verifySpeculativeTypeInfo(ifStatement.SpanStart, PublicNullableFlowState.MaybeNull);
+            verifySpeculativeTypeInfo(ifStatement.Statement.SpanStart, PublicNullableFlowState.NotNull);
+
+            verifySpeculativeTypeInfo(conditionalAccessExpression.SpanStart, PublicNullableFlowState.MaybeNull);
+            verifySpeculativeTypeInfo(conditionalAccessExpression.WhenNotNull.SpanStart, PublicNullableFlowState.NotNull);
+
+            verifySpeculativeTypeInfo(ternary.WhenTrue.SpanStart, PublicNullableFlowState.MaybeNull);
+            verifySpeculativeTypeInfo(ternary.WhenFalse.SpanStart, PublicNullableFlowState.NotNull);
+
+            void verifySpeculativeTypeInfo(int position, PublicNullableFlowState expectedFlowState)
+            {
+                var specTypeInfo = model.GetSpeculativeTypeInfo(position, newReference, SpeculativeBindingOption.BindAsExpression);
+                Assert.Equal(expectedFlowState, specTypeInfo.Nullability.FlowState);
+                specTypeInfo = model.GetSpeculativeTypeInfo(position, newCoalesce, SpeculativeBindingOption.BindAsExpression);
+                Assert.Equal(PublicNullableFlowState.NotNull, specTypeInfo.Nullability.FlowState);
+            }
+        }
+
         [Fact, WorkItem(45398, "https://github.com/dotnet/roslyn/issues/45398")]
         public void VarInLambda_GetTypeInfo()
         {

--- a/src/Compilers/Core/Portable/Collections/SmallDictionary.cs
+++ b/src/Compilers/Core/Portable/Collections/SmallDictionary.cs
@@ -736,9 +736,9 @@ hasBucket:
 
         public struct Enumerator
         {
-            private readonly Stack<AvlNode> _stack;
+            private readonly Stack<AvlNode>? _stack;
             private Node? _next;
-            private Node _current;
+            private Node? _current;
 
             public Enumerator(SmallDictionary<K, V> dict)
                 : this()
@@ -761,7 +761,7 @@ hasBucket:
                 }
             }
 
-            public KeyValuePair<K, V> Current => new KeyValuePair<K, V>(_current.Key, _current.Value);
+            public KeyValuePair<K, V> Current => new KeyValuePair<K, V>(_current!.Key, _current!.Value);
 
             public bool MoveNext()
             {
@@ -791,7 +791,7 @@ hasBucket:
             {
                 if (child != null)
                 {
-                    _stack.Push(child);
+                    _stack!.Push(child);
                 }
             }
         }

--- a/src/Compilers/Core/Portable/Collections/SmallDictionary.cs
+++ b/src/Compilers/Core/Portable/Collections/SmallDictionary.cs
@@ -515,9 +515,9 @@ hasBucket:
 
             public struct Enumerator
             {
-                private readonly Stack<AvlNode> _stack;
+                private readonly Stack<AvlNode>? _stack;
                 private Node? _next;
-                private Node _current;
+                private Node? _current;
 
                 public Enumerator(SmallDictionary<K, V> dict)
                     : this()
@@ -538,7 +538,7 @@ hasBucket:
                     }
                 }
 
-                public K Current => _current.Key;
+                public K Current => _current!.Key;
 
                 public bool MoveNext()
                 {
@@ -568,7 +568,7 @@ hasBucket:
                 {
                     if (child != null)
                     {
-                        _stack.Push(child);
+                        _stack!.Push(child);
                     }
                 }
             }
@@ -630,9 +630,9 @@ hasBucket:
 
             public struct Enumerator
             {
-                private readonly Stack<AvlNode> _stack;
+                private readonly Stack<AvlNode>? _stack;
                 private Node? _next;
-                private Node _current;
+                private Node? _current;
 
                 public Enumerator(SmallDictionary<K, V> dict)
                     : this()
@@ -655,7 +655,7 @@ hasBucket:
                     }
                 }
 
-                public V Current => _current.Value;
+                public V Current => _current!.Value;
 
                 public bool MoveNext()
                 {
@@ -685,7 +685,7 @@ hasBucket:
                 {
                     if (child != null)
                     {
-                        _stack.Push(child);
+                        _stack!.Push(child);
                     }
                 }
             }

--- a/src/Workspaces/Core/Portable/Rename/ConflictResolution.cs
+++ b/src/Workspaces/Core/Portable/Rename/ConflictResolution.cs
@@ -37,8 +37,10 @@ namespace Microsoft.CodeAnalysis.Rename
         private readonly ImmutableDictionary<DocumentId, ImmutableArray<ComplexifiedSpan>> _documentToComplexifiedSpansMap;
         private readonly ImmutableDictionary<DocumentId, ImmutableArray<RelatedLocation>> _documentToRelatedLocationsMap;
 
+#nullable disable warnings
         public ConflictResolution(string errorMessage) : this()
             => ErrorMessage = errorMessage;
+#nullable enable warnings
 
         public ConflictResolution(
             Solution oldSolution,

